README.md
cataloniaframework_v1_sample_website
====================================

Sample www.cataloniaframework.com


doc/changelog.txt
CataloniaWeb Changelog
======================

2015-08-23
Updated main vied to show support for PHP 7

2014-03-27
Added graphic for stats from today in Dashboard

Added visits to the website from today

2014-03-25
Increased messages to 25

2014-02-03
Carles Mateo
Added detected Ip to the Dashboard

2014-02-02
Carles Mateo
Added number of visits yesterday, and past yesterday to Dashboard

2014-02-01
Carles Mateo
Dashboard and full login working

Added login and logout links

Added USER_NAME to session

Added USER_NAME when logged, up to the right

Added View of last 2 contact messages in the dashboard after login

2014-01-29
Carles Mateo:
Added cron server_stats.php that tracks the CPU used every 5 minutes

2014-01-28
Carles Mateo:
Updated Documentation

2014-01-27
Carles Mateo:
Updated Documentation

2014-01-26
Carles Mateo:
Added Login controllers, models, Data Object formlogin

Added USER_ID, USER_ACCOUNT, USER_ROLE to Session as constants.

Added dashboard, for login in only

------------------------------------------------------------------

2014-01-07
Carles Mateo:
Moved visit logging to CommonRequests::logRequest

Improved table visits

Table new_install created

Fixed a warning when there were no errors in Contact Form ($st_view_vars['s_embed_javascript'] did not exist)

2014-01-06
Carles Mateo:
Updated core files to v.1.1.013

2013-12-28
Carles Mateo:
Switch the web to work with mysqli driver (was still using mysql)

Improvements in index.php and Db class (got the last Framework's version)

Finished Contact Controller, Model, Views, Translations

2013-12-26
Carles Mateo:
Added contact controller, contact model, ans saving messages.

Requests::getClientIp

Expand the manual and general improvements

2013-12-25
Carles Mateo:
Bootstrap improved to support multi-development dirs

2013-12-23
Carles Mateo:
Fixed a NOTICE when there was no HTTP_REFERER

New method Requests::getHttpReferer

2013-12-22
Carles Mateo:
Changed CSS for links on site-nav Menu

Updated from Core Framework: Added support for right bar at getSectionUrl

Added section Download, Manual

Changed title from the app

Added default parameters to Actions: $s_url = '', $st_params = Array(), $st_params_url = Array(), $o_db = null

2013-09-15 v.1.1.010 Released

catfwcore/bootstrap.php
<?php
/**
 * Creator:      Carles Mateo
 * Date:         2013-02-07 12:53
 * Last Updater: Carles Mateo
 * Last Updated: 2014-01-26 18:22
 * Filename:     bootstrap.php
 * Description:
 */

require_once 'requests.class.php';  // For Multi-Development environments in development config
use CataloniaFramework\Requests as Requests;

require_once '../config/general.php';
// Db is required in general
use CataloniaFramework\Db as Db;

// Define Autoload
require_once 'autoload.php';

// Register our CatFw Autoload
spl_autoload_register('catfw_autoload');

require_once CATFW_CORE_ROOT.'strings.class.php';
use CataloniaFramework\Strings as Strings;
require_once CATFW_CORE_ROOT.'section.class.php';
use CataloniaFramework\Section as Section;
require_once CATFW_CORE_ROOT.'core.class.php';
use CataloniaFramework\Core as Core;
require_once CATFW_CORE_ROOT.'core.php';
require_once CATFW_CORE_ROOT.'controllerbase.class.php';
use CataloniaFramework\ControllerBase as ControllerBase;
require_once CATFW_CORE_ROOT.'modelbase.class.php';
require_once CATFW_CORE_ROOT.'views.class.php';
use CataloniaFramework\Views as Views;
require_once CATFW_CORE_ROOT.'datetime.class.php';
use CataloniaFramework\DateTime as DateTime;
require_once CATFW_CORE_ROOT.'customexceptions.class.php';
require_once CATFW_CORE_ROOT.'navigation.class.php';
use CataloniaFramework\Navigation as Navigation;
require_once CATFW_CORE_ROOT.'section.class.php';
require_once CATFW_CORE_ROOT.'menu.class.php';
use CataloniaFramework\Menu as Menu;
require_once CATFW_CORE_ROOT.'translations.class.php';
use CataloniaFramework\Translations as Translations;
require_once CATFW_CORE_ROOT.'numbers.class.php';
use CataloniaFramework\Numbers as Numbers;
require_once CATFW_CORE_ROOT.'currency.class.php';
use CataloniaFramework\Currency as Currency;
// Register common to all pages, f.e. user vars
require_once CUSTOM_INIT_ROOT.'commonrequests.class.php';
use CataloniaFramework\CommonRequests as CommonRequests;
require_once CATFW_CORE_ROOT.'form.class.php';
use CataloniaFramework\Form as Form;
require_once CATFW_CORE_ROOT.'session.class.php';
use CataloniaFramework\Session as Session;
require_once CATFW_CORE_ROOT.'security.class.php';
use CataloniaFramework\Security as Security;
require_once CATFW_CORE_ROOT.'file.class.php';
use CataloniaFramework\File as File;
require_once CATFW_CORE_ROOT.'graphics.class.php';
use CataloniaFramework\Graphics as Graphics;

$s_user_language = LANGUAGE_DEFAULT;

$o_db = new Db($st_server_config['database']);

$b_is_custom_section = false;

$s_params = Requests::getParamStringGET('params');
$st_params = Strings::getParamsFromURL($s_params);

define('REQUESTED_PATH', $s_params);

require_once CUSTOM_INIT_ROOT.'customprebootstrap.php';

CommonRequests::initSession($o_db);

// Define user custom hardcoded URLs
CommonRequests::registerURLS();

if (Navigation::isURLCustom(REQUESTED_PATH) == true) {
    // We use a special method to load Custom Content
    $s_controller = 'Index';
    $s_action = 'CATFW_getCustomContent';

    if (!Core::loadController($s_controller)) {
        echo getErrorView(Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND, Views::$st_ERROR_MESSAGES[Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND]);
        Core::end();
    }

    Core::loadController($s_controller);

    $s_controller_with_namespace = '\\CataloniaFramework\\'.$s_controller;
    $o_controller = new $s_controller_with_namespace;

} else {
    // Is not custom url, so we will get the class and method to execute

    // This is a special index for the case of the basic domain call without any controller specified
    $s_controller = 'Index';
    // Default Action name
    $s_action = 'Index';

    if (MULTILANG == true) {
        if (isset($st_params[0])) {
            if (Core::isLanguageActive($st_params[0]) == false) {
                Core::redirectUserToUrl('/'.$s_user_language.'/');
                Core::end();
            }
            // The URL language is correct
            $s_user_language = $st_params[0];
            define('USER_LANGUAGE', $s_user_language);
            unset($st_params[0]);
        } else {
            // Got http://domain/ but the right thing would be http://domain/lang/ , so redirect
            Core::redirectUserToUrl('/'.$s_user_language.'/');
            Core::end();
        }

        // Register user sections
        CommonRequests::registerSections($o_db);

        if (isset($st_params[1]) && !empty($st_params[1])) {

            $st_possible_section = null;
            if (isset($st_params[2]) && !empty($st_params[2])) {
                // Check first for Section of two levels. E.g. /en/dashboard/main
                $st_possible_section = Section::getSectionInfoByPath('/'.$s_user_language.'/'.$st_params[1].'/'.$st_params[2]);
                if ($st_possible_section !== null) {
                    // Section Found
                    $b_is_custom_section = true;
                    $s_controller   = $st_possible_section['controller'];
                    $s_action       = $st_possible_section['action'];
                }
            }

            if ($st_possible_section == null) {
                // Check for matching Section of one level. E.g. /en/dashboard
                $st_possible_section = Section::getSectionInfoByPath('/'.$s_user_language.'/'.$st_params[1]);

                if ($st_possible_section !== null) {
                    // Section Found
                    $b_is_custom_section = true;
                    $s_controller   = $st_possible_section['controller'];
                    $s_action       = $st_possible_section['action'];
                } else {
                    // Common structure /lang/controller/action
                    if (isset($st_params[1]) && !empty($st_params[1])) {
                        $s_controller = Strings::getSanitizedControllerName($st_params[1]);
                        unset($st_params[1]);
                    }
                    if (isset($st_params[2]) && !empty($st_params[2])) {
                        $s_action = Strings::getSanitizedControllerName($st_params[2]);
                        unset($st_params[2]);
                    }
                }

            }

        }

        Section::registerSection('index', '/'.$s_user_language.'/');
    } else {

        define('USER_LANGUAGE', LANGUAGE_DEFAULT);
        if (isset($st_params[0]) && !empty($st_params[0])) {
            $st_possible_section = Section::getSectionInfoByPath($st_params[0]);
            if ($st_possible_section !== null) {
                // Section Found
                $b_is_custom_section = true;
                $s_controller   = $st_possible_section['controller'];
                $s_action       = $st_possible_section['action'];
            } else {
                // Common structure /controller/action
                if (isset($st_params[0]) && !empty($st_params[0])) {
                    $s_controller = Strings::getSanitizedControllerName($st_params[0]);
                    unset($st_params[0]);
                }
                if (isset($st_params[1]) && !empty($st_params[1])) {
                    $s_action = Strings::getSanitizedControllerName($st_params[1]);
                    unset($st_params[1]);
                }
            }
        }

        Section::registerSection('index', '/', 'Index', 'Index');
    }

    // Check security for controller and name
    if (!Core::isValidName($s_controller) || !Core::isValidName($s_action) ||
        !Core::loadController($s_controller)) {
        // Invalid name of controller or action
        echo getErrorView(Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND, Views::$st_ERROR_MESSAGES[Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND]);
        Core::end();
    }

    // Load the Controller Class into memory
    $s_controller_with_namespace = '\\CataloniaFramework\\'.$s_controller;
    $o_controller = new $s_controller_with_namespace;

    // For example: actionIndex
    $s_action = 'action'.$s_action;

    if (!method_exists($o_controller, $s_action)) {
        // Invalid name of controller or action
        echo getErrorView(Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND, Views::$st_ERROR_MESSAGES[Views::ERROR_CONTROLLER_OR_ACTION_NOT_FOUND]);
        Core::end();
    }

}

define('CONTROLLER', $s_controller);
define('ACTION', $s_action);

// Create parameter list
$st_params = array_values($st_params);
$st_params_url = Array();
for ($i_count=0; $i_count<=count($st_params); $i_count = $i_count+2) {
    if (isset($st_params[$i_count+1])) {
        $st_params_url[$st_params[$i_count]] = $st_params[$i_count +1];
    }
}

// Load Additional custom content
CommonRequests::registerUserVars($o_db);

// Log the visit
CommonRequests::logRequest($o_db);

require_once CUSTOM_INIT_ROOT.'custompostbootstrap.php';
